### PR TITLE
[fix] torch.nn.functional.embedding -> padding_idx behavior

### PR DESCRIPTION
**Pull request by kshitij12345**
Reference #46585 

Fix for second snippet in the mentioned issue.
```python
predefined_weights = torch.rand(10, 3)
result = torch.nn.functional.embedding(torch.LongTensor([1,2,0]), predefined_weights, padding_idx=0)
```

### CODE REVIEW

File: aten/src/ATen/native/Embedding.cpp [hunk 1]
```diff
@@ -19,14 +19,19 @@ Tensor embedding(const Tensor & weight, const Tensor & indices,
 
   // TODO: use tensor.index() after improving perf
   if (indices.dim() == 1) {
-    return weight.index_select(0, indices);
+    auto out = weight.index_select(0, indices);
+    out.masked_fill_((indices == padding_idx).reshape({-1, 1}), 0);
```
Comment (albanD) [inline aten/src/ATen/native/Embedding.cpp:23]: Out of curiosity, what is the perf implication of this change? The Embedding layer is a fairly perf-critical piece of code AFAIK

Comment (SsnL) [reply aten/src/ATen/native/Embedding.cpp:23]: it is easy to branch conditioned on `padding_idx > 0`. maybe should do that

Comment (kshitij12345) [reply aten/src/ATen/native/Embedding.cpp:23]: Ah. That makes sense. Thanks @SsnL

**Comment by kshitij12345** [comment]
@albanD Gentle Ping.

With branching on `padding_idx>=0`, do you think we still need to benchmark it once?

**Comment by albanD** [comment]
I think it would be nice to have an idea of what the perf drop is yes.
A simple check with `%timeit` for CPU and manual run on cuda with proper synchronization should be enough to get an idea.

**Comment by kshitij12345** [comment]
@albanD 

With some update to the reference script.
```python
import operator_benchmark as op_bench
import torch
import numpy
from pt import configs

"""EmbeddingBag Operator Benchmark"""

class EmbeddingBenchmark(op_bench.TorchBenchmarkBase):
    def init(self, vocab, dim, input_size, padding, device):
        numpy.random.seed((1 << 32) - 1)
        self.weight = torch.randn(vocab, dim, device=device)
        self.input = torch.tensor(numpy.random.randint(0, vocab, input_size), device=device).long()

        if padding is not None:
            padding_mask = torch.rand(self.input.shape) > 0.5
            self.input[padding_mask] = padding

        self.padding = padding
        self.set_module_name('embedding')

    def forward(self):
        return torch.nn.functional.embedding(self.input, self.weight, padding_idx=self.padding)

op_bench.generate_pt_test(configs.embedding_short_configs, EmbeddingBenchmark)

if __name__ == "__main__":
    op_bench.benchmark_runner.main()
```

Relevant Config
```python
embedding_short_configs = op_bench.cross_product_configs(
    vocab=[10000, 20000],
    dim=[64, 128],
    padding=[None, 2],
    input_size=[32, 48, 64],
    device=['cpu', 'cuda'],
    tags=['short']
)
```

Attached is the output of running the above code.

[embedding.txt](https://github.com/pytorch/pytorch/files/5447422/embedding.txt)

Let me know if the benchmark code looks good. 
Also do let me know, we should add that code in this PR?

**Comment by albanD** [comment]
That looks good. Do you have the same timings without this change to be able to compare?

**Comment by kshitij12345** [comment]
Following is the benchmark without the change.

[embedding_pre_change.txt](https://github.com/pytorch/pytorch/files/5449448/embedding_pre_change.txt)

**Review by albanD — APPROVED** [review]
Thanks for taking the time to report all the timings. 
There is a small change but not too bad. Definitely acceptable to have the correct behavior.

**Pull request closed without merging**